runner/BenchRunner.go

```diff
@@ -0,0 +1,186 @@
+// Copyright © 2019 Aqua Security Software Ltd. <info@aquasec.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runner
+
+import (
+	"archive/tar"
+	"errors"
+	"fmt"
+	"github.com/aquasecurity/bench-common/check"
+	"github.com/aquasecurity/bench-common/util"
+)
+
+type BenchRunner struct {
+	mConfigYaml []byte
+
+	//optional
+	mDefinitions []string
+
+	//this file used in FileSearch directive,
+	// instead walking thru real FS
+	mTarHeaders []tar.Header
+
+	//This directive is used by FileSearch and
+	//TextSearch to restrict search folder boundary
+	//i.e. only files in this folder are candidates for lookup
+	mPathBoundary string
+
+	//This variable used to determine either to run the shell command
+	//in yaml 'audit' (i.e. shell cmd)  or 'action' attributes
+	mIsActionTest bool
+
+	mControls *check.Controls
+
+	mCheckList string
+}
+
+func New(configYaml []byte) (runner *BenchRunner) {
+
+	ctx := new(BenchRunner)
+	ctx.mConfigYaml = configYaml
+	return ctx
+}
+
+func (ctx *BenchRunner) WithConstrains(constrains []string) *BenchRunner {
+	ctx.mDefinitions = constrains
+	return ctx
+}
+
+func (ctx *BenchRunner) WithTarHeaders(tarHeaders []tar.Header) *BenchRunner {
+	ctx.mTarHeaders = tarHeaders
+	return ctx
+}
+
+func (ctx *BenchRunner) WithWorkSpace(pathBoundary string) *BenchRunner {
+	ctx.mPathBoundary = pathBoundary
+	return ctx
+}
+
+func (ctx *BenchRunner) WithAction(isAction bool) *BenchRunner {
+	ctx.mIsActionTest = isAction
+	return ctx
+}
+
+func (ctx *BenchRunner) WithCheckList(checkList string) *BenchRunner {
+	ctx.mCheckList = checkList
+	return ctx
+}
+func (ctx *BenchRunner) Build() (*BenchRunner, error) {
+	// validate
+	if ctx.mConfigYaml == nil {
+		return nil, errors.New("ERROR empty yaml")
+	}
+
+	// try to parse the file and get controls
+	var err error
+	err = ctx.createControls()
+	if err != nil {
+		return nil, err
+	}
+	return ctx, nil
+}
+
+func (ctx *BenchRunner) createControls() (err error) {
+	ctx.mControls, err = check.NewControls(ctx.mConfigYaml).WithIsAction(ctx.mIsActionTest).
```

Comment:
This sort of functional chaining is quite fun, and I suppose it works so I'm not going to make you remove it, but it feels kinda overkill to me given that all of the "WithX" functions just assign field values. Personally I'd have kept it simple and just assigned the fields directly into the struct. "Clear is better than clever" as the [proverb ](https://go-proverbs.github.io/) goes.